src/semidiscretization.jl

```diff
@@ -147,18 +147,12 @@ end
 # The entropy/energy of the Svärd-Kalisch and Serre-Green-Naghdi equations
 # takes the whole `q` for every point in space since it requires
 # the derivative of the velocity `v_x`.
-function integrate_quantity(func::Union{typeof(energy_total_modified),
-                                        typeof(entropy_modified)}, q,
-                            semi::Semidiscretization)
-    quantity = func(q, semi.equations, semi.cache)
-    integrate(quantity, semi)
-end
-
 function integrate_quantity!(quantity,
                              func::Union{typeof(energy_total_modified),
                                          typeof(entropy_modified)}, q,
                              semi::Semidiscretization)
-    integrate_quantity(func, q, semi)
+    energy_total_modified!(quantity, q, semi.equations, semi.cache)
```

Comment:
Here, I hardcoce `energy_total_modified!`, which works because `energy_total_modified` and `entropy_modified` are doing the same thing, but which is certainly not nice. I would like to use something like `func!`. Can we achieve this somehow?

Reply:
Maybe something like 

```julia
inplace_version(::typeof(energy_total_modified)) = energy_total_modified!
inplace_version(::typeof(entropy_modified)) = entropy_modified!
```

Reply:
Looks good. Ideally, I would not need to implement this for every function. I hacked the following together:

```julia
julia> inplace_version(f) = getfield(@__MODULE__, Symbol(string(nameof(f)) * "!"))

julia> inplace_version(energy_total_modified)
energy_total_modified! (generic function with 4 methods)

julia> inplace_version(entropy_modified)
entropy_modified! (generic function with 1 method)
```
Do you think it's too hacky to put into DispersiveShallowWater.jl?

Reply:
Is it type stable?

Reply:
Checking with `@code_warntype` gives:

```julia
julia> inplace_version(f) = getfield(@__MODULE__, Symbol(string(nameof(f)) * "!"))
inplace_version (generic function with 1 method)

julia> foo(x) = 2
foo (generic function with 1 method)

julia> foo!(x) = 3
foo! (generic function with 1 method)

julia> @code_warntype inplace_version(foo)
MethodInstance for inplace_version(::typeof(foo))
  from inplace_version(f) @ Main REPL[1]:1
Arguments
  #self#::Core.Const(inplace_version)
  f::Core.Const(foo)
Body::typeof(foo!)
1 ─ %1 = Main.nameof(f)::Core.Const(:foo)
│   %2 = Main.string(%1)::Core.Const("foo")
│   %3 = (%2 * "!")::Core.Const("foo!")
│   %4 = Main.Symbol(%3)::Core.Const(:foo!)
│   %5 = Main.getfield(Main, %4)::Core.Const(foo!)
└──      return %5
```
The `Core.Const` is fine, isn't it? Or should I check something else?

Reply:
Yeah, looks good 👍

Reply:
This was not yet implemented. I added `inplace_version` in eb6e3c1a3edce07f10c0ec645a4923d338f16e56.